Persephone/Controllers/QueueViewController.swift

```diff
@@ -0,0 +1,183 @@
+  }
+
+  func setQueueIcon(_ state: mpd_state) {
+    switch state {
+    case MPD_STATE_PLAY:
+      self.queueIcon = NSImage(named: NSImage.Name("playButton"))
+    case MPD_STATE_PAUSE:
+      self.queueIcon = NSImage(named: NSImage.Name("pauseButton"))
+    default:
+      self.queueIcon = nil
+    }
+  }
+
+  func setQueuePos(oldSongRowPos: Int, newSongRowPos: Int) {
+    if oldSongRowPos > 0 {
+      guard let oldSongRow = queueView.rowView(atRow: oldSongRowPos, makeIfNecessary: true)
+        else { return }
+      guard let oldSongTitleCell = oldSongRow.view(atColumn: 0) as? NSTableCellView
+        else { return }
+
+      setRowFont(rowView: oldSongRow, font: systemFontRegular)
+      oldSongTitleCell.imageView?.image = nil
+    }
+
+    guard let songRow = queueView.rowView(atRow: newSongRowPos, makeIfNecessary: true)
+      else { return }
+    guard let newSongTitleCell = songRow.view(atColumn: 0) as? NSTableCellView
+      else { return }
+
+    setRowFont(rowView: songRow, font: systemFontBold)
+    newSongTitleCell.imageView?.image = self.queueIcon
+  }
+
+  func setRowFont(rowView: NSTableRowView, font: NSFont) {
+    guard let songTitleCell = rowView.view(atColumn: 0) as? NSTableCellView
+      else { return }
+    guard let songArtistCell = rowView.view(atColumn: 1) as? NSTableCellView
+      else { return }
+
+    songTitleCell.textField?.font = font
+    songArtistCell.textField?.font = font
+  }
+
+  func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
+    return queue.count + 1
+  }
+
+  func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
+    return false
+  }
+
+  func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
+    if index > 0 {
+      return SongItem(song: queue[index - 1], queuePos: index - 1)
+    } else {
+      return ""
+    }
+  }
+
+  func outlineView(_ outlineView: NSOutlineView, viewFor tableColumn: NSTableColumn?, item: Any) -> NSView? {
+    if let songItem = item as? SongItem {
+      switch tableColumn?.identifier.rawValue {
+      case "songTitleColumn":
+        let cellView = outlineView.makeView(
+          withIdentifier: NSUserInterfaceItemIdentifier("songTitleCell"),
+          owner: self
+        ) as! NSTableCellView
+
+        cellView.textField?.stringValue = songItem.song.getTag(MPD_TAG_TITLE)
+
+        return cellView
+      case "songArtistColumn":
+        let cellView = outlineView.makeView(
+          withIdentifier: NSUserInterfaceItemIdentifier("songArtistCell"),
+          owner: self
+        ) as! NSTableCellView
+
+        cellView.textField?.stringValue = songItem.song.getTag(MPD_TAG_ARTIST)
+
+        return cellView
+      default:
+        return nil
+      }
+    } else {
+      if tableColumn?.identifier.rawValue == "songTitleColumn" {
+        let cellView = outlineView.makeView(
+          withIdentifier: NSUserInterfaceItemIdentifier("queueHeadingCell"),
```

Comment:
What do you think about replacing all these strings ("songTitleCell", "songArtistCell", etc.) with either enum values or a series of constants (for compile-time safety)?

Reply:
Right — and an option here would be to place your constant inside an extension on the `NSUserInterfaceItemIdentifier` type:

```swift
extension NSUserInterfaceItemIdentifier {
  static let queueHeading = NSUserInterfaceItemIdentifier(“queueHeadingCell”)
}
```

Which you can then use in a nice and concise way:

```swift
let cellView = outlineView.makeView(withIdentifier: .queueHeading, owner: self)
```

Reply:
@louis-antonopoulos Can you give me an example of what you mean with the enum approach? I'm struggling to get it right.